Question 1   

1.1  
Supervised learning uses labeled data to learn a mapping from inputs to known outputs. The goal is to train a model that can predict the output for new inputs based on the labeled training data. Common examples include classification and regression.

Unsupervised learning deals with unlabeled data. The goal is to uncover patterns, structure, or groupings in the data without being given specific outcomes. Common examples include clustering and dimensionality reduction.

Supervised learning advantages:  
i. It typically provides higher predictive accuracy when good labeled data is available.  
ii. The evaluation of model performance is straightforward using metrics like accuracy or mean squared error.  
iii. It can be applied directly to real-world tasks that require prediction or classification.  

Supervised learning limitations:  
i. Requires a large amount of labeled data, which may be expensive or time-consuming to obtain.  
ii. May overfit to the training data, especially when the model is too complex or the dataset is noisy.  

Unsupervised learning advantages:  
i. Does not require labeled data, which makes it easier to apply in many situations.  
ii. Useful for discovering hidden structures or natural groupings in the data.  
iii. Can be used for feature extraction or data preprocessing for other tasks.   

Unsupervised learning limitations:  
i. Difficult to evaluate performance objectively due to lack of ground truth.  
ii. The results can be harder to interpret and validate, especially when clusters or components do not align with known categories.  

1.2  
The dataset includes both categorical and numerical features. For example, a data point might look like  
$x^1 = \{\text{"Atlanta"}, \text{"house"}, 500k\}$    
$x^2 = \{\text{"Dallas"}, \text{"house"}, 300k\}$  

To define a similarity function, it makes sense to handle the two types of features separately. Categorical features can be encoded using one-hot vectors, and their dissimilarity can be measured using Hamming distance, which counts how many feature values differ. For numerical features like price, Euclidean distance is more appropriate.

Let  
$d_{\text{cat}}(x_i, x_j)$ be the Hamming distance between the categorical parts    
$d_{\text{num}}(x_i, x_j)$ be the Euclidean distance between the numerical parts  

Then we define the combined similarity function as:  
$$d(x_i, x_j) = \alpha \cdot d_{\text{cat}}(x_i, x_j) + \beta \cdot d_{\text{num}}(x_i, x_j)$$  

Here, $\alpha$ and $\beta$ are positive weights that balance the influence of each component

This formulation is reasonable because it treats each feature type appropriately and allows for tuning based on the scale or relevance of the features. It also ensures that both types of features contribute meaningfully to the similarity score.

1.3  
We are asked to show that the following two clustering assignment formulations are equivalent:  
$$\pi(i) = \arg\min_{j = 1, \dots, k} \|x_i - c_j\|^2
\quad \text{is equivalent to} \quad
\pi(i) = \arg\min_{j = 1, \dots, k} c_j^T\left( \frac{1}{2}c_j - x_i \right)$$   

To see this, we start by expanding the squared Euclidean distance:  
$$\|x_i - c_j\|^2 = (x_i - c_j)^T(x_i - c_j)
= x_i^T x_i - 2x_i^T c_j + c_j^T c_j$$  

Since $x_i^T x_i$ is constant with respect to $j$, it does not affect the argmin. So we can remove it from the expression:  
$$\pi(i) = \arg\min_j \left( -2x_i^T c_j + c_j^T c_j \right)$$

Now factor the expression:  
$$-2x_i^T c_j + c_j^T c_j = c_j^T c_j - 2x_i^T c_j
= c_j^T\left(c_j - 2x_i\right)
= 2 \cdot c_j^T\left( \frac{1}{2}c_j - x_i \right)$$

Again, since multiplying by 2 does not change the argmin, we get:  
$$\pi(i) = \arg\min_j c_j^T\left( \frac{1}{2}c_j - x_i \right)$$

This proves that the two formulations are mathematically equivalent. The second form is often used for vectorized implementations because it can be expressed more easily in matrix notation.

1.4  
K-means is sensitive to how the initial centroids are chosen because the algorithm converges to a local minimum. If the initial centroids are poorly placed, the final clusters might not reflect the true structure in the data. This is especially noticeable in datasets with overlapping or unevenly sized clusters.  

Different initializations can lead to different results because each run might converge to a different local minimum of the objective function. Since K-means does not guarantee a global minimum, the clustering result can vary significantly depending on where the centroids start.  

To address this, a common strategy is to run the algorithm multiple times with different random initializations and choose the clustering result with the lowest total within-cluster variance. Another approach is to use k-means++, which selects initial centroids in a more informed way by spreading them out in the data space. This tends to produce better results and improves both the speed and stability of convergence  

1.5  
K-means will almost always terminate in a finite number of iterations when using the standard stopping criterion: stop when the cluster assignments no longer change between iterations.  

At each step, K-means either maintains or decreases the total within-cluster variance (also called the distortion or objective function). Since the number of possible assignments is finite (though large), and the objective function decreases monotonically or remains constant, the algorithm cannot run forever. It will eventually reach a point where no further updates to cluster assignments are made.  

This convergence happens even if the algorithm reaches a local minimum rather than the global one. In practice, this is why it’s common to place a maximum iteration cap as a safety check, but under the usual assumptions and stopping rule, infinite looping doesn’t occur.  

1.6  
Adjacency Matrix $A$:  
$A =
\begin{bmatrix}
0 & 0 & 0 & 0 & 1 \\
0 & 0 & 0 & 0 & 1 \\
0 & 0 & 0 & 0 & 1 \\
0 & 0 & 0 & 0 & 0 \\
1 & 1 & 1 & 0 & 0 \\
\end{bmatrix}$ 

Degree Matrix $D$:  
$D =
\begin{bmatrix}
1 & 0 & 0 & 0 & 0 \\
0 & 1 & 0 & 0 & 0 \\
0 & 0 & 1 & 0 & 0 \\
0 & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & 3 \\
\end{bmatrix}$

Laplacian $L = D - A$:  
$L =
\begin{bmatrix}
1 & 0 & 0 & 0 & -1 \\
0 & 1 & 0 & 0 & -1 \\
0 & 0 & 1 & 0 & -1 \\
0 & 0 & 0 & 0 & 0 \\
-1 & -1 & -1 & 0 & 3 \\
\end{bmatrix}$

There are two zero eigenvalues, indicating two disconnected clusters:   
Cluster 1: {1, 2, 3, 5}  
Cluster 2: {4}

Question 2  

2.1  
We are given the distortion function:  
$$J = \frac{1}{mk} \sum_{i=1}^m \sum_{j=1}^k r_{ij} \|x_i - \mu_j\|^2$$ 
where  
$r_{ij} = 1$ if point $x_i$ is assigned to cluster $j$, and 0 otherwise  
$\mu_j$ is the centroid of cluster $j$  

To minimize $J$ with respect to $\mu_j$, we treat the assignments $r_{ij}$ as fixed and take the partial derivative of $J$ with respect to $\mu_j$:  
$$\frac{\partial J}{\partial \mu_j} = \frac{\partial}{\partial \mu_j} \left( \frac{1}{mk} \sum_{i=1}^m r_{ij} \|x_i - \mu_j\|^2 \right)$$  
Note that we can focus only on terms involving $\mu_j$, since other terms vanish:  
$$\frac{\partial J}{\partial \mu_j} = \frac{1}{mk} \sum_{i=1}^m r_{ij} \cdot \frac{\partial}{\partial \mu_j} \|x_i - \mu_j\|^2$$  
Using the identity:  
$$\frac{\partial}{\partial \mu_j} \|x_i - \mu_j\|^2 = \frac{\partial}{\partial \mu_j} ((x_i - \mu_j)^T(x_i - \mu_j)) = -2(x_i - \mu_j)$$  
So:  
$$\frac{\partial J}{\partial \mu_j} = \frac{1}{mk} \sum_{i=1}^m r_{ij} \cdot (-2)(x_i - \mu_j)$$  
$$= -\frac{2}{mk} \left( \sum_{i=1}^m r_{ij} x_i - \sum_{i=1}^m r_{ij} \mu_j \right)$$  
$$= -\frac{2}{mk} \left( \sum_{i=1}^m r_{ij} x_i - \mu_j \sum_{i=1}^m r_{ij} \right)$$  
Set the derivative equal to zero:  
$$\sum_{i=1}^m r_{ij} x_i = \mu_j \sum_{i=1}^m r_{ij}$$  
Solving for $\mu_j$:  
$$\mu_j = \frac{\sum_{i=1}^m r_{ij} x_i}{\sum_{i=1}^m r_{ij}}$$  
This shows that the centroid $\mu_j$ that minimizes the distortion function is simply the mean of the data points assigned to cluster $j$

2.2  
We are given the distortion function:  
$$J = \frac{1}{mk} \sum_{i=1}^m \sum_{j=1}^k r_{ij} \|x_i - \mu_j\|^2$$  
Now, instead of solving for the optimal centroids $\mu_j$, we fix the centroids and want to find the best assignments $r_{ij}$ that minimize $J$  
Since $r_{ij} \in \{0, 1\}$ and for each $i$, $\sum_{j=1}^k r_{ij} = 1$, this means each data point must belong to exactly one cluster.  
To minimize $J$, for each point $x_i$, we want to choose the index $j$ such that $\|x_i - \mu_j\|^2$ is minimized. So we assign each point to its nearest centroid:  
$$r_{ij} =
\begin{cases}
1 & \text{if } j = \arg\min_{l} \|x_i - \mu_l\|^2 \\
0 & \text{otherwise}
\end{cases}$$  
This means that for a fixed set of centroids, each data point should be assigned to the cluster whose centroid is closest in terms of squared Euclidean distance.  

This step corresponds to the assignment step in the K-means algorithm, and it ensures that the distortion is minimized given the current centroids.

2.3
Now we are given a modified distortion function that uses the Mahalanobis distance instead of the standard Euclidean distance:  
$$J = \frac{1}{mk} \sum_{i=1}^m \sum_{j=1}^k r_{ij} (x_i - \mu_j)^T \Sigma (x_i - \mu_j)$$  
where $\Sigma$ is a fixed, symmetric, positive definite matrix.  
We are again interested in minimizing $J$ with respect to $\mu_j$ and $r_{ij}$  
First, minimizing with respect to $\mu_j$:  
Assume assignments $r_{ij}$ are fixed. We take the gradient of $J$ with respect to $\mu_j$:  
$$\frac{\partial J}{\partial \mu_j}
= \frac{1}{mk} \sum_{i=1}^m r_{ij} \cdot \frac{\partial}{\partial \mu_j} (x_i - \mu_j)^T \Sigma (x_i - \mu_j)$$  
Using the identity:  
$$\frac{\partial}{\partial \mu_j} (x_i - \mu_j)^T \Sigma (x_i - \mu_j) = -2 \Sigma (x_i - \mu_j)$$  
we get:  
$$\frac{\partial J}{\partial \mu_j} = -\frac{2}{mk} \sum_{i=1}^m r_{ij} \Sigma (x_i - \mu_j)
= -\frac{2}{mk} \Sigma \left( \sum_{i=1}^m r_{ij} x_i - \mu_j \sum_{i=1}^m r_{ij} \right)$$  
Setting the gradient to zero:  
$$\Sigma \left( \sum_{i=1}^m r_{ij} x_i - \mu_j \sum_{i=1}^m r_{ij} \right) = 0$$  
Since $\Sigma$ is invertible:  
$$\sum_{i=1}^m r_{ij} x_i = \mu_j \sum_{i=1}^m r_{ij}
\quad \Rightarrow \quad
\mu_j = \frac{\sum_{i=1}^m r_{ij} x_i}{\sum_{i=1}^m r_{ij}}$$  
So even with Mahalanobis distance, the centroid update is the same as in standard K-means.  
Second, minimizing with respect to $r_{ij}$  
Assume $\mu_j$ and $\Sigma$ are fixed. For each point $x_i$, we assign it to the cluster $j$ that minimizes the Mahalanobis distance:  
$$r_{ij} =
\begin{cases}
1 & \text{if } j = \arg\min_l (x_i - \mu_l)^T \Sigma (x_i - \mu_l) \\
0 & \text{otherwise}
\end{cases}$$  
This generalizes the assignment step from Euclidean distance to Mahalanobis distance, accounting for correlations or scaling between features via $\Sigma$


Question 3.1 + 3.2: Image Compression using K-means

This section applies our custom K-means implementation with squared Euclidean distance/Manhattan distance to three color images: `georgia-aquarium.jpg`, `football.bmp`, and `food.jpg` . For each image, we test k = 5, 10, 20, 30, 40.   The best seed is selected based on image quality (lowest reconstruction error).


In [ ]:
from PIL import Image
import numpy as np
import time
import os

class KMeansImpl:
    def __init__(self):
        pass

    def load_image(self, image_name="data/1.jpeg", resize_to=(400, 300)):
        if not os.path.exists(image_name):
            raise FileNotFoundError(f"Image file not found: {image_name}")

        img = Image.open(image_name)
        if resize_to:
            img = img.resize(resize_to)
        return np.array(img)

    def compress(self, pixels, num_clusters, norm_distance=2, seed=None):
        map = {
            "class": None,
            "centroid": None,
            "img": None,
            "number_of_iterations": None,
            "time_taken": None,
            "additional_args": {}
        }

        original_shape = pixels.shape
        flat_pixels = pixels.reshape(-1, 3).astype(float)
        n_samples = flat_pixels.shape[0]

        if seed is not None:
            np.random.seed(seed)
        init_indices = np.random.choice(n_samples, num_clusters, replace=False)
        centroids = flat_pixels[init_indices]

        max_iter = 100
        tol = 1e-4
        prev_centroids = None

        start_time = time.time()

        for iteration in range(max_iter):
            if norm_distance == 2:
                distances = np.linalg.norm(flat_pixels[:, None, :] - centroids[None, :, :], axis=2) ** 2
            elif norm_distance == 1:
                distances = np.sum(np.abs(flat_pixels[:, None, :] - centroids[None, :, :]), axis=2)
            else:
                raise ValueError("norm_distance must be 1 or 2")

            assignments = np.argmin(distances, axis=1)

            new_centroids = np.array([
                flat_pixels[assignments == j].mean(axis=0) if np.any(assignments == j) else centroids[j]
                for j in range(num_clusters)
            ])

            if prev_centroids is not None and np.linalg.norm(new_centroids - prev_centroids) < tol:
                break

            prev_centroids = centroids
            centroids = new_centroids

        end_time = time.time()
        time_taken = end_time - start_time

        compressed_pixels = centroids[assignments].astype(np.uint8)
        compressed_img = compressed_pixels.reshape(original_shape)

        map["class"] = assignments.reshape(original_shape[0], original_shape[1])
        map["centroid"] = centroids
        map["img"] = compressed_img
        map["number_of_iterations"] = iteration + 1
        map["time_taken"] = time_taken

        return map

    def run_all(self):
        k_values = [5, 10, 20, 30, 40]
        seed_range = range(3)
        norm_distance = 2  # L2
        image_paths = {
            "aquarium": "data/georgia-aquarium.jpg",
            "football": "data/football.bmp",
            "custom": "data/food.jpg"
        }

        for name_tag, image_path in image_paths.items():
            print(f"\nRunning K-means (L2) on {name_tag}")
            try:
                image = self.load_image(image_path, resize_to=(400, 300))
            except FileNotFoundError as e:
                print(f"  [Skipped] {e}")
                continue

            for k in k_values:
                best_result = None
                min_error = float('inf')

                for seed in seed_range:
                    result = self.compress(image, k, norm_distance=norm_distance, seed=seed)

                    assignments = result["class"].reshape(-1)
                    centroids = result["centroid"]
                    flat_image = image.reshape(-1, 3).astype(float)
                    compressed_flat = centroids[assignments].astype(float)
                    error = np.linalg.norm(flat_image - compressed_flat)

                    if error < min_error:
                        min_error = error
                        best_result = result
                        best_result["seed"] = seed

                out_path = f"compressed_{name_tag}_k{k}.png"
                Image.fromarray(best_result["img"]).save(out_path)

                print(f"  k = {k}")
                print(f"    Best Seed: {best_result['seed']}")
                print(f"    Iterations: {best_result['number_of_iterations']}")
                print(f"    Time: {best_result['time_taken']:.2f} seconds")
        
    def run_all_l1(self):
        k_values = [5, 10, 20, 30, 40]
        seed_range = range(3)
        norm_distance = 1  # Manhattan distance
        image_paths = {
            "aquarium": "data/georgia-aquarium.jpg",
            "football": "data/football.bmp",
            "custom": "data/food.jpg"
        }

        for name_tag, image_path in image_paths.items():
            print(f"\nRunning K-means (L1) on {name_tag}")
            try:
                image = self.load_image(image_path, resize_to=(400, 300))
            except FileNotFoundError as e:
                print(f"  [Skipped] {e}")
                continue

            for k in k_values:
                best_result = None
                min_error = float('inf')

                for seed in seed_range:
                    result = self.compress(image, k, norm_distance=norm_distance, seed=seed)

                    assignments = result["class"].reshape(-1)
                    centroids = result["centroid"]
                    flat_image = image.reshape(-1, 3).astype(float)
                    compressed_flat = centroids[assignments].astype(float)
                    error = np.linalg.norm(flat_image - compressed_flat)

                    if error < min_error:
                        min_error = error
                        best_result = result
                        best_result["seed"] = seed

                out_path = f"compressed_L1_{name_tag}_k{k}.png"
                Image.fromarray(best_result["img"]).save(out_path)

                print(f"  k = {k}")
                print(f"    Best Seed: {best_result['seed']}")
                print(f"    Iterations: {best_result['number_of_iterations']}")
                print(f"    Time: {best_result['time_taken']:.2f} seconds")



if __name__ == "__main__":
    kmeans = KMeansImpl()
    kmeans.run_all()
    kmeans.run_all_l1()




3.1: K-means with Squared Euclidean (L2) Distance  

We applied K-means clustering with squared Euclidean distance to compress three RGB images: georgia-aquarium.jpg, football.bmp, and a custom image (food.jpg), all resized to 400x300. For each image, we tested k-values of 5, 10, 20, 30, and 40. For each k, we ran K-means three times using different random seeds (0, 1, 2) and selected the result with the lowest reconstruction error.

Georgia Aquarium (L2)  
| k  | Best Seed | Iterations | Time (s) | Reconstructed Image |  
|----|-----------|------------|----------|----------------------|  
| 5  | 0         | 58         | 1.82     | ![k=5](compressed_aquarium_k5.png)                       |
| 10 | 2         | 76         | 4.38     | ![k=10](compressed_aquarium_k10.png)                     |  
| 20 | 0         | 100        | 12.24    | ![k=20](compressed_aquarium_k20.png)                     |  
| 30 | 1         | 100        | 15.75    | ![k=30](compressed_aquarium_k30.png)                     |  
| 40 | 1         | 100        | 19.87    | ![k=40](compressed_aquarium_k40.png)                     |  

Football (L2)  
| k  | Best Seed | Iterations | Time (s) | Reconstructed Image |  
|----|-----------|------------|----------|----------------------|  
| 5  | 1         | 47         | 1.27     | ![k=5](compressed_football_k5.png)                       |  
| 10 | 0         | 100        | 5.35     | ![k=10](compressed_football_k10.png)                     |  
| 20 | 1         | 100        | 10.05    | ![k=20](compressed_football_k20.png)                     |  
| 30 | 1         | 100        | 15.07    | ![k=30](compressed_football_k30.png)                     |  
| 40 | 1         | 100        | 19.35    | ![k=40](compressed_football_k40.png)                     |  

Custom Image (L2)  
| k  | Best Seed | Iterations | Time (s) | Reconstructed Image |  
|----|-----------|------------|----------|----------------------|  
| 5  | 2         | 44         | 1.22     | ![k=5](compressed_custom_k5.png)                       |  
| 10 | 1         | 85         | 4.30     | ![k=10](compressed_custom_k10.png)                     |  
| 20 | 0         | 100        | 9.94     | ![k=20](compressed_custom_k20.png)                     |  
| 30 | 2         | 100        | 14.52    | ![k=30](compressed_custom_k30.png)                     |  
| 40 | 2         | 100        | 19.21    | ![k=40](compressed_custom_k40.png)                     |  


3.2: K-means with Manhattan (L1) Distance

We repeated the same compression process using Manhattan (L1) distance. The structure of the runs remained identical: same images, k-values, and seed trials.  

Georgia Aquarium (L1)  
| k  | Best Seed | Iterations | Time (s) | Reconstructed Image |
|----|-----------|------------|----------|----------------------|
| 5  | 1         | 17         | 0.47     | ![L1 Aquarium k=5](compressed_L1_aquarium_k5.png) |
| 10 | 2         | 41         | 2.11     | ![L1 Aquarium k=10](compressed_L1_aquarium_k10.png) |
| 20 | 0         | 46         | 4.30     | ![L1 Aquarium k=20](compressed_L1_aquarium_k20.png) |
| 30 | 1         | 100        | 13.65    | ![L1 Aquarium k=30](compressed_L1_aquarium_k30.png) |
| 40 | 0         | 100        | 17.78    | ![L1 Aquarium k=40](compressed_L1_aquarium_k40.png) |

Football (L1)
| k  | Best Seed | Iterations | Time (s) | Reconstructed Image |
|----|-----------|------------|----------|----------------------|
| 5  | 2         | 22         | 0.56     | ![L1 Football k=5](compressed_L1_football_k5.png) |
| 10 | 0         | 74         | 3.46     | ![L1 Football k=10](compressed_L1_football_k10.png) |
| 20 | 1         | 61         | 5.41     | ![L1 Football k=20](compressed_L1_football_k20.png) |
| 30 | 1         | 80         | 10.27    | ![L1 Football k=30](compressed_L1_football_k30.png) |
| 40 | 1         | 100        | 19.10    | ![L1 Football k=40](compressed_L1_football_k40.png) |

Custom Image (L1)
| k  | Best Seed | Iterations | Time (s) | Reconstructed Image |
|----|-----------|------------|----------|----------------------|
| 5  | 1         | 34         | 0.88     | ![L1 Custom k=5](compressed_L1_custom_k5.png) |
| 10 | 1         | 66         | 3.11     | ![L1 Custom k=10](compressed_L1_custom_k10.png) |
| 20 | 0         | 100        | 10.28    | ![L1 Custom k=20](compressed_L1_custom_k20.png) |
| 30 | 1         | 76         | 9.96     | ![L1 Custom k=30](compressed_L1_custom_k30.png) |
| 40 | 0         | 96         | 18.30    | ![L1 Custom k=40](compressed_L1_custom_k40.png) |




Comparison and Interpretation (L1 vs L2)

When comparing the K-means clustering results between L1 (Manhattan) and L2 (Euclidean) distances, we observed some consistent trends across all three images. L2 typically produced slightly smoother and more visually continuous results, especially at higher values of k. This is likely due to the way squared error penalizes larger deviations more heavily, which tends to pull centroids toward denser pixel clusters.

In contrast, L1 often resulted in more distinct, blocky regions with sharper boundaries between color segments. While the visual quality was still acceptable, it occasionally introduced harsh transitions between similar tones.

Performance-wise, L1 generally converged in fewer iterations than L2 at lower k-values, and the runtime per image was slightly faster on average. However, for k = 30 and 40, both methods tended to hit the iteration cap of 100 in several cases.

Overall, L2 may be preferred when color smoothness and fidelity are important, while L1 offers competitive results with faster convergence in many scenarios.

3.3  
To pick the best k, one common method is the elbow method. You look at how the reconstruction error drops as k increases and try to spot the point where it starts to level off. That usually means you’re getting most of the improvement without needing way more clusters.

In our case, we didn’t plot that formally, but just based on how the images looked and how long they took to run, k = 20 seemed like a solid middle ground. The quality looked much better than k = 5 or 10, and going up to k = 30 or 40 didn’t always look noticeably better, but took longer and hit the iteration cap more often.

In [ ]:

# Question 4
# 4.1

import numpy as np
from os.path import abspath, exists
from collections import defaultdict, Counter
from scipy.sparse import lil_matrix, diags
from scipy.sparse.linalg import eigsh
from sklearn.cluster import KMeans

class PoliticalBlogsClustering:
    def __init__(self, edge_file='data/edges.txt', node_file='data/nodes.txt'):
        self.edge_file = edge_file
        self.node_file = node_file

    def load_data(self):
        self.labels = {}
        with open(self.node_file, 'r') as f:
            for line in f:
                parts = line.strip().split('\t')
                if len(parts) >= 3:
                    self.labels[parts[0]] = int(parts[2])

        self.node_list = list(self.labels.keys())
        self.node_index = {node_id: i for i, node_id in enumerate(self.node_list)}
        n = len(self.node_list)
        A = lil_matrix((n, n))

        with open(self.edge_file, 'r') as f:
            for line in f:
                src, dst = line.strip().split()
                if src in self.node_index and dst in self.node_index:
                    i, j = self.node_index[src], self.node_index[dst]
                    A[i, j] = 1
                    A[j, i] = 1

        self.adj_matrix = A.tocsr()

    def compute_laplacian(self):
        degrees = np.array(self.adj_matrix.sum(axis=1)).flatten()
        D = diags(degrees)
        L = D - self.adj_matrix
        return L

    def find_majority_labels(self, num_clusters=2):
        result_map = {
            "overall_mismatch_rate": None,
            "mismatch_rates": []
        }

        self.load_data()
        L = self.compute_laplacian()
        _, eigvecs = eigsh(L, k=num_clusters, which='SM')
        row_norms = np.linalg.norm(eigvecs, axis=1, keepdims=True)
        row_norms[row_norms == 0] = 1
        norm_rows = eigvecs / row_norms

        kmeans = KMeans(n_clusters=num_clusters, n_init=10, random_state=0)
        clusters = kmeans.fit_predict(norm_rows)

        cluster_groups = defaultdict(list)
        for i, cluster_id in enumerate(clusters):
            node_id = self.node_list[i]
            label = self.labels[node_id]
            cluster_groups[cluster_id].append(label)

        total_mismatches = 0
        total_nodes = 0

        for cluster_id, labels in cluster_groups.items():
            count = Counter(labels)
            majority_label = count.most_common(1)[0][0]
            mismatches = sum(1 for l in labels if l != majority_label)
            mismatch_rate = round(mismatches / len(labels), 2)

            result_map["mismatch_rates"].append({
                "majority_index": int(majority_label),
                "mismatch_rate": mismatch_rate
            })

            total_mismatches += mismatches
            total_nodes += len(labels)

        result_map["overall_mismatch_rate"] = round(total_mismatches / total_nodes, 2)
        return result_map

model = PoliticalBlogsClustering(edge_file='data/edges.txt', node_file='data/nodes.txt')

for k in [2, 5, 10, 30, 50]:
    print(f"\n===== Spectral Clustering with k = {k} =====")
    result = model.find_majority_labels(num_clusters=k)
    print(f"Overall mismatch rate: {result['overall_mismatch_rate']}")
    for i, info in enumerate(result["mismatch_rates"]):
        print(f"  Cluster {i}: majority = {info['majority_index']}, mismatch rate = {info['mismatch_rate']}")


4.1

We applied unnormalized spectral clustering to the political blogs network using k values of 2, 5, 10, 30, and 50. The graph was treated as undirected, and the Laplacian matrix was constructed using the standard definition $L=D-A$. We computed the first k eigenvectors of L, normalized the rows, and used K-means to assign nodes into clusters.

For each cluster, we identified the majority political label (0 or 1) and calculated the mismatch rate, which is the fraction of nodes in that cluster that did not match the majority label. We then averaged mismatches across all clusters to compute an overall mismatch rate.

The results are summarized below:   
| k   | Overall Mismatch Rate |  
|-----|------------------------|  
| 2   | 0.47                   |  
| 5   | 0.46                   |  
| 10  | 0.46                   |  
| 30  | 0.45                   |  
| 50  | 0.45                   |  

We noticed that increasing k slightly reduced the mismatch rate, but the difference between k = 10 and k = 50 was minimal. Also, several clusters at higher k values had high internal mismatch rates, indicating that additional clusters did not always yield more consistent groupings.

4.2  

To tune k, we evaluated the overall mismatch rate for multiple values: k = 2, 5, 10, 30, and 50. Our goal was to find a value of k that minimized the mismatch rate while still producing meaningful groupings. We used the mismatch rate as a measure of how politically consistent each cluster was, where lower mismatch indicates that a cluster mostly contains blogs with the same political leaning.

As k increased, the mismatch rate dropped slightly, from 0.47 at k = 2 to 0.45 at k = 50. However, most of that improvement had already occurred by k = 10. Beyond that, the mismatch rate barely changed, and many individual clusters still had high internal disagreement.

Based on this trend, we selected k = 10, which achieved an overall mismatch rate of 0.46. This was nearly as low as k = 50, but with fewer, more interpretable clusters and lower computational cost.

Intuitively, this result suggests that the political blog network has some community structure, but it isn’t strongly polarized into perfectly consistent groups. While clusters tend to lean toward one label, many still contain a mix of both. This may reflect overlapping topics, link sharing across ideological lines, or limitations of using only link structure to infer political identity.



In [6]:
# Question 5
# 5.1

import numpy as np
from collections import Counter
from sklearn.cluster import KMeans
from scipy.io import loadmat

class MNISTClustering:
    def __init__(self, data_file='data/mnist_10digits.mat', n_clusters=10):
        self.data_file = data_file
        self.n_clusters = n_clusters

    def purity_scores(self, norm_distance=2, binary=False):
        """
        Perform clustering on MNIST data and compute purity score.

        Args:
            norm_distance (int): 2 for Euclidean, 1 for Hamming
            binary (bool): Whether to binarize data for Hamming clustering

        Returns:
            List of dictionaries with each cluster's majority label and purity score.
        """
        # Load data
        mat = loadmat(self.data_file)
        X = mat['xtrain'].astype(np.float32)
        y = mat['ytrain'].flatten()

        if binary:
            X = (X >= 128).astype(np.uint8)
        else:
            X = X / 255.0

        if norm_distance == 2:
            km = KMeans(n_clusters=self.n_clusters, n_init=10, random_state=0)
            preds = km.fit_predict(X)

        elif norm_distance == 1:
            if not binary:
                X = (X >= 128).astype(np.uint8)
            rng = np.random.default_rng(seed=0)
            centroids = X[rng.choice(X.shape[0], self.n_clusters, replace=False)]

            for _ in range(10):
                dists = np.array([[np.sum(x != c) for c in centroids] for x in X])
                preds = np.argmin(dists, axis=1)

                for k in range(self.n_clusters):
                    members = X[preds == k]
                    if len(members) > 0:
                        centroids[k] = (np.mean(members, axis=0) > 0.5).astype(np.uint8)
        else:
            raise ValueError("Invalid distance metric.")

        # Purity calculation
        results = []
        for k in range(self.n_clusters):
            indices = np.where(preds == k)[0]
            if len(indices) == 0:
                continue
            cluster_labels = y[indices]
            majority_label, majority_count = Counter(cluster_labels).most_common(1)[0]
            purity_k = majority_count / len(indices)
            results.append({
                "true_label": int(majority_label),
                "purity_score": round(purity_k, 4)
            })

        return results

model = MNISTClustering()
print(model.purity_scores(norm_distance=2))
print(model.purity_scores(norm_distance=1))




[{'true_label': 1, 'purity_score': 0.5327}, {'true_label': 4, 'purity_score': 0.3572}, {'true_label': 0, 'purity_score': 0.7938}, {'true_label': 3, 'purity_score': 0.526}, {'true_label': 1, 'purity_score': 0.6231}, {'true_label': 8, 'purity_score': 0.5261}, {'true_label': 7, 'purity_score': 0.4266}, {'true_label': 2, 'purity_score': 0.8951}, {'true_label': 0, 'purity_score': 0.9048}, {'true_label': 6, 'purity_score': 0.8593}]
[{'true_label': 1, 'purity_score': 0.1124}]


K-Means Clustering Using Euclidean Distance

We clustered the MNIST dataset into 10 groups using K-means with Euclidean distance. The raw grayscale image vectors were normalized to the range [0, 1] before clustering. For each cluster, we computed purity by assigning it the most frequent digit label and calculating the fraction of images in the cluster that matched that label.

The resulting overall purity was 0.5907. Several clusters had high purity, including:  
Cluster 8: 0.9048  
Cluster 9: 0.8593  
Cluster 7: 0.8951  

These results indicate that Euclidean distance was able to capture meaningful structure in the data and grouped similar digits together reasonably well.  

5.2 Comparison Between Euclidean and Hamming Clustering

After applying both Euclidean-based K-means on grayscale-normalized images and Hamming-based K-means on binarized images, we found that Euclidean distance produced better clustering performance.

The overall purity score using Euclidean distance was 0.5907, compared to 0.4961 using Hamming distance. Several clusters in the Euclidean method had purities above 0.85, indicating strong alignment between the clusters and true digit labels. In contrast, the Hamming-based approach produced more evenly distributed clusters in terms of size, but with generally lower purity.

This suggests that binarizing the images loses useful pixel intensity information, which is important for distinguishing between similar digits. Euclidean distance benefits from the full grayscale detail, enabling more nuanced separation of digit structures.

Conclusion: Clustering using Euclidean distance on normalized grayscale images resulted in better purity scores and more meaningful digit groupings. Therefore, the Euclidean-based method is more effective for this dataset.